http/src/v0/refs.rs

```diff
@@ -138,32 +138,253 @@ async fn refs_paths<T: IpfsTypes>(
+                Ok(MaybeResolved::NeedToLoadMore(lookup)) => lookup,
+                Ok(MaybeResolved::Found(cid)) => {
+                    current = cid;
+                    continue;
+                }
+                Ok(MaybeResolved::NotFound) => {
+                    if needle == "Data" && path.len() == 0 {
+                        // /dag/resolve needs to "resolve through" a dag-pb node down to the "just
+                        // data" even though we do not need to extract it ... however this might be
+                        // good to just filter with refs, as no refs of such path exist
+                        //
+                        // testing with go-ipfs 0.5 reveals that dag resolve only follows links
+                        // which are actually present in the dag-pb, not numeric links like Links/5
+                        // or links/5, even if such are present in the `dag get` output.
+                        //
+                        // comment on this special casing: there cannot be any other such
+                        // special case as the Links do not work like Data so while this is not
+                        // pretty, it's not terrible.
+                        let data = ipfs::unixfs::ll::dagpb::node_data(&data)
+                            .expect("already parsed once, second time cannot fail")
+                            .unwrap_or_default();
+                        return Ok((
+                            current,
+                            Loaded::Ipld(Ipld::Bytes(data.to_vec())),
+                            vec![needle],
+                        ));
+                    }
+                    let e = WalkFailed::from(path::WalkFailed::UnmatchedNamedLink(needle));
+                    return Err(WalkError::from((e, current)));
+                }
+                Err(e) => return Err(WalkError::from((WalkFailed::from(e), current))),
+            };
+
+            loop {
+                let (next, _) = lookup.pending_links();
+
+                // need to take ownership in order to enrich the error, next is invalidaded on
+                // lookup.continue_walk.
+                let next = next.to_owned();
+
+                let Block { data, .. } = match ipfs.get_block(&next).await {
+                    Ok(block) => block,
+                    Err(e) => return Err(WalkError::from((WalkFailed::from(e), next))),
+                };
+
+                match lookup.continue_walk(&data, &mut cache) {
+                    Ok(MaybeResolved::NeedToLoadMore(next)) => lookup = next,
+                    Ok(MaybeResolved::Found(cid)) => {
+                        current = cid;
+                        break;
+                    }
+                    Ok(MaybeResolved::NotFound) => {
+                        if needle == "Data" && path.len() == 0 {
+                            // see remarks in the earlier MaybeResolve::NotFound branch
```

Comment:
this block seems to be the same as the one this comment refers to; perhaps we can use a function (on `data`, `needle` and `path`) to avoid repeating ourselves?

Reply:
Not a bad idea :)